src/main/java/com/limechain/trie/structure/Vacant.java

```diff
@@ -0,0 +1,482 @@
+package com.limechain.trie.structure;
+
+import com.limechain.trie.structure.nibble.Nibble;
+import com.limechain.trie.structure.nibble.Nibbles;
+import lombok.AllArgsConstructor;
+import org.apache.commons.lang3.ArrayUtils;
+import org.javatuples.Pair;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.Arrays;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.stream.IntStream;
+
+public final class Vacant<T> extends Entry<T> {
+    /**
+     * Full key of the node to insert.
+     */
+    private final Nibbles key;
+
+    /**
+     * The closest known ancestor. Will become the parent of any newly-inserted node at this vacant spot.
+     */
+    @Nullable
+    private final Integer closestAncestorIndex;
+
+    Vacant(TrieStructure<T> trieStructure, Nibbles key, @Nullable Integer closestAncestorIndex) {
+        super(trieStructure);
+        this.key = key;
+        this.closestAncestorIndex = closestAncestorIndex;
+    }
+
+    /**
+     * Prepares the operation of creating the node in question.
+     *
+     * @implNote This method analyzes the trie to prepare for the operation, but doesn't actually perform any insertion.
+     * To perform the insertion, use the returned {@link PrepareInsert}.
+     * @return a {@link PrepareInsert} object containing all the data necessary to actually perform the insertion.
+     */
+    public PrepareInsert<T> prepareInsert() {
+        // Retrieve what will be the parent node after we insert the new node,
+        // not taking branching into account yet.
+        // If not null, contains its index and number of nibbles in its key.
+        Pair<Integer, Integer> futureParent = null;
+        futureParentInit: {
+            // If the trie is empty
+            if (this.trieStructure.rootIndex == null) {
+                // ... but we somehow found an ancestor...
+                if (this.closestAncestorIndex != null) {
+                    // UNREACHABLE STATE
+                    throw new IllegalStateException("Unreachable state reached: ancestor found in an empty trie.");
+                }
+
+                // ... and no ancestor has been rightfully found, simply insert one node (will effectively be the root)
+                // This is kind of a special case that we handle by returning early.
+                return new PrepareInsert.One<>(
+                    this.trieStructure,
+                    null,
+                    key.copy(),
+                    new Integer[16]
```

Comment:
I thought about this, but honestly, in the context of looking through this class and what it does (collecting data to insert nodes), the plain number was a much more clear indicator of "ah yes, those are some node's children" (since I know they're exactly 16) than any variable name I could think of. The literal `16` in and of itself seemed like a fundamental enough name (again, in this context 😄).

What would a reasonable name for this magic number be? `TRIE_NODE_CHILDREN_COUNT`?

Reply:
Yea, that seems like a fine name.